contrib/epee/include/net/levin_protocol_handler_async.h

```diff
@@ -190,115 +190,102 @@ class async_protocol_handler
 
   struct invoke_response_handler_base
   {
-    virtual bool handle(int res, const epee::span<const uint8_t> buff, connection_context& context)=0;
-    virtual bool is_timer_started() const=0;
-    virtual void cancel()=0;
-    virtual bool cancel_timer()=0;
-    virtual void reset_timer()=0;
-  };
-  template <class callback_t>
-  struct anvoke_handler: invoke_response_handler_base
-  {
-    anvoke_handler(const callback_t& cb, uint64_t timeout,  async_protocol_handler& con, int command)
-      :m_cb(cb), m_timeout(timeout), m_con(con), m_timer(con.m_pservice_endpoint->get_io_service()), m_timer_started(false),
-      m_cancel_timer_called(false), m_timer_cancelled(false), m_command(command)
-    {
-      if(m_con.start_outer_call())
-      {
-        MDEBUG(con.get_context_ref() << "anvoke_handler, timeout: " << timeout);
-        m_timer.expires_from_now(boost::posix_time::milliseconds(timeout));
-        m_timer.async_wait([&con, command, cb, timeout](const boost::system::error_code& ec)
-        {
-          if(ec == boost::asio::error::operation_aborted)
-            return;
-          MINFO(con.get_context_ref() << "Timeout on invoke operation happened, command: " << command << " timeout: " << timeout);
-          epee::span<const uint8_t> fake;
-          cb(LEVIN_ERROR_CONNECTION_TIMEDOUT, fake, con.get_context_ref());
-          con.close();
-          con.finish_outer_call();
-        });
-        m_timer_started = true;
-      }
-    }
-    virtual ~anvoke_handler()
-    {}
-    callback_t m_cb;
-    async_protocol_handler& m_con;
+  protected:
     boost::asio::deadline_timer m_timer;
-    bool m_timer_started;
-    bool m_cancel_timer_called;
-    bool m_timer_cancelled;
-    uint64_t m_timeout;
-    int m_command;
-    virtual bool handle(int res, const epee::span<const uint8_t> buff, typename async_protocol_handler::connection_context& context)
+    async_protocol_handler& m_con;
+    const uint64_t m_timeout;
+    const int m_command;
+
+    virtual void do_handle(int res, const epee::span<const uint8_t> buff)=0;
```

Comment:
Same here: It would save an extra 8-byte copy to make buff a const reference instead of pass-by-value

Reply:
Hmm. Typically this is bad practice because the caller is required to provide a memory address for the span. On most (modern) ABIs this often results in less efficient code (because at most call-sites there is no existing span on the stack). In this instance, the function is invoked from one spot, and that chunk of code probably/maybe has a span on the stack already. 

I'm tempted to leave it because its dependent on the setup of the existing code. Could be persuaded differently, I guess.

Reply:
I was just reminded by this [blog post(https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/) that by value is likely preferred here.

The callee doesn't have to generate indirect memory dereferences (the values are directly in the register), and it doesn't have to worry about the span being aliased (better optimizations in some cases).

I doubt either is going to matter much in this situation, but its a series of tradeoffs to save one move instruction in the optimal case.

Reply:
Yeah fair enough. The only reason I mentioned it was because it was just passing it through to `do_handle`. But either way, the profiling isn't worth effort. Nevermind this comment.